lib/snippets/kubernetes-deploy

```diff
@@ -66,99 +200,455 @@ class KubernetesDeploy
+      }
+    end
+
+    def group_name
+      type + "s"
+    end
+
+    def run_kubectl(*args)
+      raise FatalDeploymentError, "Namespace missing for namespaced command" if namespace.blank?
+      args = args.unshift("kubectl").push("--namespace=#{namespace}")
+      KubernetesDeploy.logger.debug Shellwords.join(args)
+      out, err, st = Open3.capture3(*args)
+      KubernetesDeploy.logger.debug(out.shellescape)
+      KubernetesDeploy.logger.debug("[ERROR] #{err.shellescape}") unless st.success?
+      [out.chomp, st]
+    end
+
+    def log_status
+      STDOUT.puts "[KUBESTATUS] #{JSON.dump(status_data)}"
+    end
+  end
+
+  class ConfigMap < KubernetesResource
+    TIMEOUT = 30
+
+    def initialize(name, namespace, file)
+      @name, @namespace, @file = name, namespace, file
+    end
+
+    def sync
+      _, st = run_kubectl("get", type, @name)
+      @status = st.success? ? "Available" : "Unknown"
+      @found = st.success?
+      log_status
+    end
+
+    def deploy_succeeded?
+      exists?
+    end
+
+    def deploy_failed?
+      false
+    end
+
+    def exists?
+      @found
+    end
+  end
+
+  class PersistentVolumeClaim < KubernetesResource
+    TIMEOUT = 60 * 5
+
+    def initialize(name, namespace, file)
+      @name, @namespace, @file = name, namespace, file
+    end
+
+    def sync
+      @status, st = run_kubectl("get", type, @name, "--output=jsonpath={.status.phase}")
+      @found = st.success?
+      log_status
+    end
+
+    def deploy_succeeded?
+      @status == "Bound"
+    end
+
+    def deploy_failed?
+      @status == "Lost"
+    end
+
+    def exists?
+      @found
+    end
+  end
+
+  class Ingress < KubernetesResource
+    TIMEOUT = 30
+
+    def initialize(name, namespace, file)
+      @name, @namespace, @file = name, namespace, file
+    end
+
+    def sync
+      _, st = run_kubectl("get", type, @name)
+      @status = st.success? ? "Created" : "Unknown"
+      @found = st.success?
+      log_status
+    end
+
+    def deploy_succeeded?
+      exists?
+    end
+
+    def deploy_failed?
+      false
+    end
+
+    def exists?
+      @found
+    end
+
+    def group_name
+      "Ingresses"
+    end
+  end
+
+  class Service < KubernetesResource
+    TIMEOUT = 60 * 15
+
+    def initialize(name, namespace, file)
+      @name, @namespace, @file = name, namespace, file
+    end
+
+    def sync
+      _, st = run_kubectl("get", type, @name)
+      @found = st.success?
+      if @found
+        endpoints, st = run_kubectl("get", "endpoints", @name, "--output=jsonpath={.subsets[*].addresses[*].ip}")
+        @num_endpoints = (st.success? ? endpoints.split.length : 0)
+      else
+        @num_endpoints = 0
+      end
+      @status = "#{@num_endpoints} endpoints"
+      log_status
+    end
+
+    def deploy_succeeded?
+      @num_endpoints > 0
+    end
+
+    def deploy_failed?
+      false
+    end
+
+    def exists?
+      @found
+    end
+  end
+
+  class Pod < KubernetesResource
+    TIMEOUT = 60 * 15
```

Comment:
Totally a fair question! Please feel free to comment on things like this. It's super arbitrary. What timeout would you suggest here?

My thought was that run-once pods could theoretically be doing something that takes quite a while (e.g. a dirty `rake db:migrate` that does maintenance instead of enqueuing jobs to do maintenance), and they can always manually abort sooner. Fundamentally, these timeouts are to protect Shipit from running forever for no reason. App developers can set timeouts on their pods, and if those get hit, the deploy will hard fail immediately.